### ISSUE TITLE
Confusing doc: Eq, PartialEq, etc. are not implemented because..

### ISSUE THREAD

**Issue by ivanceras** [briansmith/webpki#107]
The [doc] says: 

> Eq, PartialEq, etc. are **not** implemented because name comparison frequently should be done case-insensitively and/or with other caveats that depend on the specific circumstances in which the comparison is done.

However, we can see that PartialEq and Eq are indeed implemented using `#[derive]` 

```rust
impl PartialEq<DNSName> for DNSName

impl Eq for DNSName
```
[doc]: https://docs.rs/webpki/0.21.0/webpki/struct.DNSName.html

**Comment by djc** [comment]
Yes, and I was pretty surprised by the fact that the `PartialEq` implementation is case insensitive. While I understand that this could technically make sense, I'm not sure it fits with the common usage of `PartialEq`, and thus it might be better to expose a separate method to do case-insensitive comparisons and leave `PartialEq` to more conservatively compare DNS names.

**Comment by briansmith** [comment]
Here's the history of this as I remember it: In linkerd2-proxy, we were using webpki's `DNSName` [sic]  (now `DnsName`) as the type of DNS names in the product. We did that because we (I) wanted to ensure we only supported DNS names that could be used in TLS. And also in the past people had requested a similar case-insensitive comparison so that they could, for example, maintain a lookup table from the DNS Name to the certificate bundle, according to the TLS/X.509 rules, which mandate case-insensitive comparison. I've been meaning to add similar comparison between `DnsNameRef`s; it's just a bit harder because we have to do more work to avoid allocation.

Keep in mind that all name comparisons in webpki are case-insensitive because that's what X.509 requires. However, now that I think about it, this is somewhat at odds with the idea that certificates are DER encoded and DER is supposed to be a canonical form of ASN.1; this means a good ASN.1 encoder should do case folding at encoding time. We should ask other X.509 implementers if they'd be willing to switch to be so strict.

**Comment by briansmith** [comment]
Regardless, this is definitely a valid bug because the documentation says `Eq` and `PartialEq` are intentionally not implemented for `DnsName` when in fact they are.

**Comment by briansmith** [comment]
@djc For your issue with Rustls, Rustls is trying to enforce something stricter than X.509/WebpPKI rules require, so you can't expect to use any X.509 certificate library's name comparison to do the stricter check you're trying to do.
